arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/ForwardDisposable.kt

```diff
@@ -0,0 +1,70 @@
+package arrow.fx.reactor.extensions
+
+import arrow.fx.internal.Platform
+import kotlinx.atomicfu.atomic
+import reactor.core.Disposable
+
+/**
+ * [ForwardDisposable] allows you to link 2 [Disposable]s.
+ *
+ * This is useful for when you want to link the cancel lifecycle of an `Emitter`
+ * to running an [Disposable] stream.
+ *
+ * This is used in `bracket`, `guaranteeCase` & `asyncF`.
+ */
+internal class ForwardDisposable {
+
+  private val state = atomic(init)
+
+  fun cancel(): Disposable {
+    fun loop(d: Disposable): Unit = state.value.let { current ->
+      when (current) {
+        is State.Empty -> if (!state.compareAndSet(current, State.Empty(listOf(d) + current.stack))) loop(d)
+        is State.Active -> {
+          state.lazySet(finished)
+          Platform.trampoline { current.token.dispose() }
+        }
+      }
+    }
+
+    return object : Disposable {
+      override fun dispose() =
+        loop(this)
+    }
+  }
+
+  fun complete(value: Disposable): Unit = state.value.let { current ->
+    when (current) {
+      is State.Active -> {
+        value.dispose()
+        throw IllegalStateException(current.toString())
+      }
+      is State.Empty -> if (current == init) {
```

Comment:
mmmmmmm `Disposable` does't have an identity. I'd use `===` to do reference comparison instead.

Reply:
Actually `EMPTY_DISPOSABLE` exists as an `empty` reference iirc.